packages/next/client/index.tsx

```diff
@@ -79,6 +80,17 @@ envConfig.setConfig({
 
 let asPath = getURL()
 
+if (gsp) {
+  // this does not decode query values only the path
+  // ensuring non-ascii values are the proper encoding
+  // and matches SSR/prerender
+  try {
+    asPath = decodeURI(asPath)
+  } catch (_) {
+    // non-fatal
+  }
+}
```

Comment:
How does this handle routes that are intentionally double encoded by the user when visiting?

Reply:
What would be a case you intentionally double encode the path? This aims to have the `asPath` match the same level of decoding as we do when generating the paths at build time as otherwise we can have a hydration mis-match

Reply:
When would this ever error? We control the gsp asPath so it should always be properly encoded, right?

BTW a hydration mismatch *is* fatal, so we can't swallow this.

Reply:
This is now updated to ensure we use the fully encoded variant while prerendering and still output the path delimiter escaped version to the filesystem. This should prevent hydration mismatch from a properly encoded URL being visited on the client but the prerendered version not being properly encoded.